src/context.rs

```diff
@@ -0,0 +1,244 @@
+use std::{
+    any::{Any, TypeId},
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap, HashSet, VecDeque},
+};
+
+use derivative::Derivative;
+
+pub trait Plugin: Any {
+    type DataContainer;
+
+    fn get_data_container() -> Self::DataContainer;
+}
+
+#[macro_export]
+macro_rules! define_plugin {
+    ($plugin:ident, $data_container:ty, $default: expr) => {
+        struct $plugin {}
+
+        impl $crate::context::Plugin for $plugin {
+            type DataContainer = $data_container;
+
+            fn get_data_container() -> Self::DataContainer {
+                $default
+            }
+        }
+    };
+}
+pub use define_plugin;
+
+pub struct PlanId {
+    pub id: u64,
+}
+
+#[derive(Derivative)]
+#[derivative(Eq, PartialEq, Debug)]
+pub struct TimedPlan {
+    pub time: f64,
+    plan_id: u64,
+    #[derivative(PartialEq = "ignore", Debug = "ignore")]
+    pub callback: Box<dyn FnOnce(&mut Context)>,
+}
+
+impl Ord for TimedPlan {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let time_ordering = self.time.partial_cmp(&other.time).unwrap().reverse();
+        if time_ordering == Ordering::Equal {
+            // Break time ties in order of plan id
+            self.plan_id.cmp(&other.plan_id).reverse()
+        } else {
+            time_ordering
+        }
+    }
+}
+
+impl PartialOrd for TimedPlan {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Debug)]
+struct PlanQueue {
+    queue: BinaryHeap<TimedPlan>,
+    invalid_set: HashSet<u64>,
+    plan_counter: u64,
+}
+
+impl PlanQueue {
+    pub fn new() -> PlanQueue {
+        PlanQueue {
+            queue: BinaryHeap::new(),
+            invalid_set: HashSet::new(),
+            plan_counter: 0,
+        }
+    }
+
+    pub fn add_plan(&mut self, time: f64, callback: impl FnOnce(&mut Context) + 'static) -> PlanId {
+        // Add plan to queue and increment counter
+        let plan_id = self.plan_counter;
+        self.queue.push(TimedPlan {
+            time,
+            plan_id,
+            callback: Box::new(callback),
+        });
+        self.plan_counter += 1;
+        PlanId { id: plan_id }
+    }
+
+    pub fn cancel_plan(&mut self, id: PlanId) {
+        self.invalid_set.insert(id.id);
```

Comment:
As we discussed live, this should be structured in some way that it will panic or at least warn if you try to cancel a plan that has already run. I think you don't want to be in a scenario where there's a bug in your logic where you do the cancel too late but it just sort of fails silently.

Reply:
Yes, good observation. I implemented this in one relatively easy way and added tests.